packages/langium/src/workspace/workspace-lock.ts

```diff
@@ -0,0 +1,129 @@
+     *
+     * Note that read actions are not allowed to modify anything in the workspace. Please use {@link write} instead.
+     */
+    read<T>(action: () => MaybePromise<T>): Promise<T>;
+
+    /**
+     * Cancels the last queued write action. All previous write actions already have been cancelled.
+     */
+    cancelWrite(): void;
+}
+
+type LockAction<T = void> = (token: CancellationToken) => MaybePromise<T>;
+
+interface LockEntry {
+    action: LockAction<unknown>;
+    deferred: Deferred<unknown>;
+    cancellationToken: CancellationToken;
+}
+
+export class DefaultWorkspaceLock implements WorkspaceLock {
+
+    private previousTokenSource = new CancellationTokenSource();
+    private writeQueue: LockEntry[] = [];
+    private readQueue: LockEntry[] = [];
+    private done = true;
+
+    write(action: (token: CancellationToken) => MaybePromise<void>): Promise<void> {
+        this.cancelWrite();
+        const tokenSource = new CancellationTokenSource();
+        this.previousTokenSource = tokenSource;
+        return this.enqueue(this.writeQueue, action, tokenSource.token);
+    }
+
+    read<T>(action: () => MaybePromise<T>): Promise<T> {
+        return this.enqueue(this.readQueue, action);
+    }
+
+    private enqueue<T = void>(queue: LockEntry[], action: LockAction<T>, cancellationToken?: CancellationToken): Promise<T> {
+        const deferred = new Deferred<unknown>();
+        const entry: LockEntry = {
+            action,
+            deferred,
+            cancellationToken: cancellationToken ?? CancellationToken.None
+        };
+        queue.push(entry);
+        this.tryPerformNextOperation();
+        return deferred.promise as Promise<T>;
+    }
+
+    private tryPerformNextOperation(): void {
+        if (this.done) {
+            this.performNextOperation();
+        }
+    }
+
+    private performNextOperation(): void {
+        if (this.writeQueue.length > 0) {
+            // Just perform the next write action
+            this.performOperation([this.writeQueue.shift()!]);
+        } else if (this.readQueue.length > 0) {
+            // Empty the read queue and perform all actions in parallel
+            const entries = this.readQueue.splice(0, this.readQueue.length);
+            this.performOperation(entries);
+        }
+    }
+
+    private performOperation(entries: LockEntry[]): void {
+        if (!this.done) {
+            throw new Error('Mutex is not ready to accept new operation');
+        }
+        this.done = false;
+        let completed = 0;
+        const goToNext = () => {
+            completed += 1;
+            // If all specified actions have been completed, we can perform the next operation
+            if (completed === entries.length) {
+                this.done = true;
+                this.performNextOperation();
+            }
+        };
+        for (const entry of entries) {
```

Comment:
Would `Promise.all(...)` work here, too?
Is there a benefit of this sequential loop?

Reply:
My thought ist

With returning the result of `Promise.all(...)` the `goToNext` construction could be skipped and `performNextOperation()` could be written like

```
private performNextOperation(): void {
        let promise: Promise;
        if (this.writeQueue.length > 0) {
            // Just perform the next write action
            promise = this.performOperation([this.writeQueue.shift()!]);
        } else if (this.readQueue.length > 0) {
            // Empty the read queue and perform all actions in parallel
            const entries = this.readQueue.splice(0, this.readQueue.length);
            promise = this.performOperation(entries);
        }
        if (promise) {
            promise.finally(() => performNextOperation());
        }
}
```

which shows the 'run to completion' characteristics better.